src/migrate-config-options.js

```diff
@@ -0,0 +1,104 @@
+'use babel'
+
+/*
+ * These migrations can take one of two forms, a direct move or a general function.
+ *
+ * Direct move:
+ *   These objects have an array of `moves`, which
+ *   are objects containing an `old` setting name and a `new` setting name.
+ *   Any existing config found in the `old` name will be moved over to (and overwrite)
+ *   the `new` key.
+ *
+ * Functions:
+ *   These have a `migrate` function, which takes the
+ *   current linter-eslint atom config as an argument, and can act on it however
+ *   it needs to.
+ */
+const migrations = [
+  {
+    added: 'January, 2018',
+    description: 'Organized config settings into sections',
+    moves: [
+      {
+        old: 'disableWhenNoEslintConfig',
+        new: 'disabling.disableWhenNoEslintConfig',
+      }, {
+        old: 'fixOnSave',
+        new: 'autofix.fixOnSave'
+      }, {
+        old: 'ignoreFixableRulesWhileTyping',
+        new: 'autofix.ignoreFixableRulesWhileTyping'
+      }, {
+        old: 'rulesToDisableWhileFixing',
+        new: 'autofix.rulesToDisableWhileFixing'
+      }, {
+        old: 'rulesToSilenceWhileTyping',
+        new: 'disabling.rulesToSilenceWhileTyping'
+      }, {
+        old: 'disableEslintIgnore',
+        new: 'advanced.disableEslintIgnore'
+      }, {
+        old: 'disableFSCache',
+        new: 'advanced.disableFSCache'
+      }, {
+        old: 'showRuleIdInMessage',
+        new: 'advanced.showRuleIdInMessage'
+      }, {
+        old: 'eslintrcPath',
+        new: 'global.eslintrcPath'
+      }, {
+        old: 'advancedLocalNodeModules',
+        new: 'advanced.advancedLocalNodeModules'
+      }, {
+        old: 'eslintRulesDirs',
+        new: 'advanced.eslintRulesDirs'
+      }, {
+        old: 'useGlobalEslint',
+        new: 'global.useGlobalEslint'
+      }, {
+        old: 'globalNodePath',
+        new: 'global.globalNodePath'
+      }
+    ]
+  },
+  {
+    added: 'September, 2017',
+    description: 'Deprecated eslintRulesDir{String} option in favor of eslintRulesDirs{Array<String>}',
+    migrate(config) {
+      const oldRulesdir = config.eslintRulesDir
+      if (oldRulesdir) {
+        const newRulesDirs = config.eslintRulesDirs
+        if (newRulesDirs.length === 0) {
+          atom.config.set('linter-eslint.eslintRulesDirs', [oldRulesdir])
+        }
+        atom.config.unset('linter-eslint.eslintRulesDir')
+      }
+    }
+  }
+]
+
+/*
+ * This function can be called when linter-eslint first activates in order to
+ * ensure that the user's settings are up-to-date with the current version of
+ * linter-eslint.  Ideally, we would call this only when upgrading to a new
+ * version.
+ */
+function migrateConfigOptions() {
+  const linterEslintConfig = atom.config.get('linter-eslint')
```

Comment:
🤔 What about only grabbing this in the `forEach` if it's not already set?

It's not quite as clean, but it _does_ save grabbing the entire settings if there are no migrations to perform.

Reply:
Based on the way this is laid out, I don't see any situation where your suggestion would somehow prevent grabbing the config. The `if` check is just so he can handle `eslintRulesDirs` differently than the straight renames.

Reply:
I'm thinking about later, when we completely deprecate and remove the current migrations. Unless we are going to keep them around forever.

Reply:
Well as long as we're messing around with it, I think this could be much more clear by increasing the symmetry between different types of migrations, as in they should _all_ be function calls, which also removes the need to check if `moves` exists. So combining that with your "maybe don't grab config", that would probably look something like this.

```
const migrations = [
  {
    added: 'January, 2018',
    description: 'Organized config settings into sections',
    migrate: (config) => {
      const moves = [ /* ... list of moves ... */ ]
      moves.forEach((move) => {
        const oldSetting = config[move.old]
        if (oldSetting !== undefined) {
          atom.config.set(move.new, oldSetting)
          atom.config.unset(oldSetting)
        }
      })
    }
  },
  {
    added: 'September, 2017'
    // ... 
  }
]

// If nothing to migrate don't grab config
const linterEslintConfig = migrations.length 
  ? atom.config.get('linter-eslint')
  : null

// If nothing to migrate, no problem that config is
// null because this is a no-op.
migrations.forEach(migration => migration.migrate(config))
```

Reply:
I'd move that inner `(move) =>` function out to it's own definition since it will likely be used elsewhere, but that does look cleaner.